### ISSUE TITLE
Help wanted: debugging scaling in static HMM model

### ISSUE THREAD

**Issue by femtomc** [probcomp/Gen.jl#289]
Hi all!

I'm trying to perform some benchmark comparisons between `Gen` and a system I setup which is very similar to `Gen` called `Jaynes`. In particular, I recently implemented the notion of "specialized call sites" in Jaynes which are close to the combinators of Gen, and provide similar performance benefits when used to indicate dependency information in a program.

[I've been working on generating a benchmark here.](https://github.com/femtomc/JaynesBenchmarks/blob/master/src/JaynesBenchmarks.jl)

I've run in to some issues with getting the static DSL to work properly, I think I'm making some stupid mistake somewhere - because the scaling with time of Gen inference is not following the correct linear trend [shown in the tutorial.](https://www.gen.dev/particle-filtering-in-gen/Particle%20Filtering%20in%20Gen)

It appears to be showing the scaling you would expect with the dynamic DSL, and I'm not sure why (if I call `load_generated_functions` at the wrong place, could this cause it?) If anyone has a bit of time, and you're willing to check out the `Gen` portion of the above code, I'd like to figure out what I'm doing wrong.

**Comment by femtomc** [comment]
Note: here's some images from my experimentation.

This is an image where I'm posting a choice map constraint for the next time step.

![benchmark_hmmpf_##gen#759](https://user-images.githubusercontent.com/34410036/88239862-67364200-cc3a-11ea-866c-ebd878cc580f.png)

This is without any constraints (i.e. unconstrained generation) - `Jaynes` is still constrained, hence the difference in log marginal likelihood of evidence.

![benchmark_hmmpf_##gen#485](https://user-images.githubusercontent.com/34410036/88240327-9dc08c80-cc3b-11ea-98c7-16ffa9b0a7a4.png)

I suspected that maybe the observations were messing something up - but it appears not. So there must be some constant overhead somewhere. I'll try and do some profiling as part of the benchmarking and post back.

**Comment by alex-lew** [comment]
> So there must be some constant overhead somewhere.

@femtomc Thanks for making these plots! This doesn't look like constant overhead to me -- it looks like asymptotic scaling is for some reason incorrect. 

I played with your code for a bit and I think the reason is that in your Gen code, you perform an O(N) operation every step, in this line:

```
return [init_state, states...]
```

This constructs a new list by iterating over all of the `states`. Note that you don't have this line in your Jaynes version of the model.

Change the line to:

```
return (init_state, states)
```

and the performance should improve significantly.

**Comment by femtomc** [comment]
Thank you @alex-lew! I forgot I was unpacking and allocating there. If I had just profiled at first I would have seen this.

I suspect this modification will show what I expect - Gen static + combinators should be faster by some margin.

**Comment by alex-lew** [comment]
> I suspect this modification will show what I expect - Gen static + combinators should be faster by some margin.

@femtomc We'll see! I'd expect a small constant-factor difference in performance, but not sure to whose advantage. The Jaynes code seems to be doing the same thing as the Gen code, but via dynamos instead of macros -- not sure what the performance differences there would be. Gen's code also uses PersistentVectors, which slow things down but can help make SMC faster when you use resampling. (Have you considered adding resampling to your benchmark? As it currently stands, you're not doing any resampling, so you're really benchmarking importance sampling, rather than what people typically think of as particle filtering. See Gen's `maybe_resample!` function.)

**Comment by femtomc** [comment]
That's exactly the implementation - except combinators are now activated as special call sites, and not explicitly called on generative functions (or Julia functions, which would be the `Jaynes` equivalent, as the dynamos can inspect non-wrapped functions). I did notice the `PersistentVector` difference - I did not know that was for resampling! Good point.

I will include that in the benchmark, thanks for the tip as well.

**Comment by femtomc** [comment]
![benchmark_hmmpf_##gen#694](https://user-images.githubusercontent.com/34410036/88297953-655d9480-ccb5-11ea-9205-6a7d228e11db.png)

Edit: this earlier comment was a mistake - while there still may be some constant overhead, one factor I did not consider is that I've turned on multi-threading for inference algorithms in `Jaynes` - hence the constant factor difference. Here's a much more telling plot.

![benchmark_hmmpf_##gen#781](https://user-images.githubusercontent.com/34410036/88323212-bfbb1d00-ccd6-11ea-9502-554e1708f4b0.png)

which is what I expected.

Out of curiosity, will `Gen` enable multi-threading at some point? This is very easy to do now partially (just add `Threads.@threads` to loops, for example).
